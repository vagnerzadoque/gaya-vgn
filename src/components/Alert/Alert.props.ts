import { IconName } from '@naturacosmeticos/natds-icons'
import React from 'react'

export interface AlertProps {
  /**
   * Specify an optional className to be added to your Alert
   */
  className?: string

  /**
   * Optional ID for testing
   */
  testID?: string

  /**
   * Parameter to define the style that will be applied
   * @default info
  */
  color?: 'info' | 'success' | 'error' | 'warning' | 'custom'

  /**
   * Parameter to define border styles
   * @default info
  */
  type?: 'contained' | 'outlined'

  /**
   * The text to be define as Heading
  */
  title?: string

  /**
   * Parameter to render a custom Icon using color as custom
  */
  customIcon?: IconName

  /**
   * Parameter to render a custom Icon using color as custom
   * @default true
  */
  showIcon?: boolean

  /**
   * The body text of component to render on the left side
  */
  children: React.ReactNode

  /**
   * Parameter to set custom background color
   * Should be a hexadecimal color like e.g. #000000
  */
  customBackgroundColor?: string

  /**
   * Parameter to set custom font color
   * Should be a hexadecimal color like e.g. #000000
  */
  customTextColor?: string

  /**
   * Parameter to set custom icon color
   * Should be a hexadecimal color like e.g. #000000
  */
  customIconColor?: string

  /**
   * Parameter to set custom border color
   * Should be a hexadecimal color like e.g. #000000
  */
  customBorderColor?: string
}
