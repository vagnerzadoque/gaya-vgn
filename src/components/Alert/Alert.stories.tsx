import React from 'react'
import type { Meta, StoryObj } from '@storybook/react'
import { Alert, AlertProps } from '.'

const componentStatus = `
> Alert allows you to display notification or warning messages on the user interface

## Properties
| Property                      | Values                                 |    Status           |
|---                            |                                     ---|                  ---|
| **variant**                   | standard                               | ✅ Available        |
| **type**                      | contained, outlined                    | ✅ Available        |
| **color**                     | info, success, error, warning, custom  | ✅ Available        |
| **title**                     | string                                 | ✅ Available        |
| **customIcon**                | icon-name                              | ✅ Available        |
| **showIcon**                  | true/false                             | ✅ Available        |
| **customBackgroundColor**     | string                                 | ✅ Available        |
| **customTextColor**           | string                                 | ✅ Available        |
| **customIconColor**           | string                                 | ✅ Available        |
| **customBorderColor**         | string                                 | ✅ Available        |


## Technical Usages Examples

`

const meta: Meta<AlertProps> = {
  title: 'Components/Alert',
  component: Alert,
  parameters: {
    componentSubtitle: '',
    docs: { description: { component: componentStatus } }
  }
};

export default meta;
type Story = StoryObj<AlertProps>;

export const Playground: Story = {
  args: {
    title: "Title",
    children: "Lorem, ipsum dolor sit amet consectetur adipisicing elit."
  },
  render: (args) => <Alert {...args} />
};

export const WithoutIcon: Story = {
  args: {
    title: "Title",
    showIcon: false,
    children: "Lorem, ipsum dolor sit amet consectetur adipisicing elit."
  },
  render: (args) => <Alert {...args} />
};

export const WithoutTitle: Story = {
  args: {
    children: "Lorem, ipsum dolor sit amet consectetur adipisicing elit."
  },
  render: (args) => <Alert {...args} />
};

export const Custom: Story = {
  args: {
    customIcon: "outlined-default-mockup",
    type: "outlined",
    color: "custom",
    customBackgroundColor: "#FF00FF",
    customBorderColor: "#FF00FF",
    customIconColor: "#FF00FF",
    customTextColor: "#FF00FF",
    children: "Lorem, ipsum dolor sit amet consectetur adipisicing elit."
  },
  render: (args) => <Alert {...args} />
};
