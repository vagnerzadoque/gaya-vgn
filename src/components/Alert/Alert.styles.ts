/* eslint-disable complexity */
import { createUseStyles } from 'react-jss'
import { Theme } from '@naturacosmeticos/natds-themes'

type VariantTypes = 'info' | 'success' | 'error' | 'warning' | 'custom'
type BorderTypes = 'contained' | 'outlined'

type StyleProps = {
  color: VariantTypes
  borderType: BorderTypes
  customBackgroundColor?: string
  customTextColor?: string
  customIconColor?: string
  customBorderColor?: string
}

const setBaseStyle = () => ({
  color, borderType, customBackgroundColor, customIconColor, customBorderColor
}: StyleProps) => {
  const variantStyle = {
    info: {
      background: '#D0EBFF',
      border: borderType === 'outlined' ? 'solid 1px #1C7ED6' : 'none'
    },
    success: {
      background: '#D3F9D8',
      border: borderType === 'outlined' ? 'solid 1px #37B24D' : 'none'
    },
    error: {
      background: '#FFE3E3',
      border: borderType === 'outlined' ? 'solid 1px #F03E3E' : 'none'
    },
    warning: {
      background: '#FFF3BF',
      border: borderType === 'outlined' ? 'solid 1px #F59F00' : 'none'
    },
    custom: {
      background: `${customBackgroundColor}29`,
      border: borderType === 'outlined' ? `solid 1px ${customBorderColor}` : 'none',
      '& > i': {
        color: customIconColor
      }
    }
  }

  return variantStyle[color]
}

const styles = createUseStyles((theme: Theme) => ({
  wrapper: (props: StyleProps) => ({
    ...setBaseStyle()(props),
    display: 'flex',
    gap: theme.spacing.tiny,
    fontFamily: `"${theme.typography.body.regular.fontFamily}"`,
    padding: theme.spacing.small,
    borderRadius: theme.alert.borderRadius
  }),
  content: {
    color: ({ color, customTextColor }) => (color === 'custom' ? customTextColor : theme.color.highEmphasis),
    '& > h6': {
      margin: `0 0 ${theme.spacing.tiny}px 0`,
      fontSize: theme.heading6.fontSize
    }
  }
}))

export default styles
