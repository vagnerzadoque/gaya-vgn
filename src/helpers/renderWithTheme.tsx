import React from 'react'
import { JssProvider, SheetsRegistry } from 'react-jss'
import themes from '@naturacosmeticos/natds-themes'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

const renderWithTheme = (component: JSX.Element) => {
  const sheets = new SheetsRegistry()

  const renderTree = render(
    <JssProvider registry={sheets}>
      <ThemeProvider theme={themes.natura.light}>
        {component}
      </ThemeProvider>
    </JssProvider>
  )

  return { styles: sheets.toString(), component: renderTree }
}

export default renderWithTheme
