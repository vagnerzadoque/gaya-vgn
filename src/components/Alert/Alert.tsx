import React from 'react'
import { IconName } from '@naturacosmeticos/natds-icons'
import Icon from '../Icon'
import styles from './Alert.styles'
import { AlertProps } from './Alert.props'

type DefaultIcons = {
  [key: string]: IconName
}

const DEFAULT_ICONS: DefaultIcons = {
  info: 'outlined-alert-info',
  success: 'outlined-alert-check',
  error: 'outlined-alert-cancel',
  warning: 'outlined-alert-warning'
}

const Alert = React.forwardRef<HTMLDivElement, AlertProps>(
  (props, ref) => {
    const {
      className = '',
      color = 'info',
      title,
      type = 'contained',
      showIcon = true,
      customIcon,
      customBackgroundColor,
      customIconColor,
      customTextColor,
      customBorderColor,
      ...rest
    } = props

    const { wrapper, content } = styles({
      color,
      borderType: type,
      customBackgroundColor,
      customIconColor,
      customTextColor,
      customBorderColor
    })

    const renderIcon = () => {
      let iconName

      if (color === 'custom' && customIcon) {
        iconName = customIcon
      }

      if (color !== 'custom') {
        iconName = DEFAULT_ICONS[color]
      }

      return iconName ? <Icon size="standard" name={iconName} color="highEmphasis" /> : null
    }

    return (
      <div
        className={`${className} ${wrapper} Alert-gaya`}
        ref={ref}
        {...rest}
      >
        {showIcon && renderIcon()}

        <div className={content}>
          {title && <h6>{title}</h6>}
          {props.children}
        </div>
      </div>
    )
  }
)

export default Alert
