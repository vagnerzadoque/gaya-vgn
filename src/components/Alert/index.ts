import Alert from './Alert'

export { Alert }
export { AlertProps } from './Alert.props'
export default Alert
