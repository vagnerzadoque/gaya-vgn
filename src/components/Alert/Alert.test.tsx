import { screen } from '@testing-library/react'
import React from 'react'
import Alert from '.'
import renderWithTheme from '../../helpers/renderWithTheme'

describe('Alert component', () => {
  it('should render component with info styles if color is not defined', () => {
    const { styles, component } = renderWithTheme(
      <Alert title="Title">Lorem, ipsum dolor sit amet consectetur adipisicing elit.</Alert>
    )

    expect(screen.getByTestId('icon-outlined-alert-info')).toBeInTheDocument()
    expect(screen.getByRole('heading')).toBeInTheDocument()
    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render component without icon if showIcon is false', () => {
    renderWithTheme(
      <Alert title="Title" showIcon={false}>Lorem, ipsum dolor sit amet consectetur adipisicing elit.</Alert>
    )

    expect(screen.queryByTestId('icon-outlined-alert-info')).not.toBeInTheDocument()
  })

  it('should render component without title if is not defined', () => {
    renderWithTheme(
      <Alert>Lorem, ipsum dolor sit amet consectetur adipisicing elit.</Alert>
    )

    expect(screen.queryByRole('heading')).not.toBeInTheDocument()
  })
})

describe('Alert colors', () => {
  it('should render component with success styles', () => {
    renderWithTheme(
      <Alert color="success">Lorem, ipsum dolor sit amet consectetur adipisicing elit.</Alert>
    )

    expect(screen.getByTestId('icon-outlined-alert-check')).toBeInTheDocument()
  })

  it('should render component with error styles', () => {
    renderWithTheme(
      <Alert color="error">Lorem, ipsum dolor sit amet consectetur adipisicing elit.</Alert>
    )

    expect(screen.getByTestId('icon-outlined-alert-cancel')).toBeInTheDocument()
  })

  it('should render component with warning styles', () => {
    renderWithTheme(
      <Alert color="warning">Lorem, ipsum dolor sit amet consectetur adipisicing elit.</Alert>
    )

    expect(screen.getByTestId('icon-outlined-alert-warning')).toBeInTheDocument()
  })
})

describe('Alert border styles', () => {
  it('should render component without borders by default', () => {
    const { styles } = renderWithTheme(
      <Alert color="success">Lorem, ipsum dolor sit amet consectetur adipisicing elit.</Alert>
    )

    expect(styles.toString()).toContain('border: none')
  })

  it('should render component with borders if type is outline', () => {
    const { styles } = renderWithTheme(
      <Alert color="success" type="outlined">Lorem, ipsum dolor sit amet consectetur adipisicing elit.</Alert>
    )

    expect(styles.toString()).toContain('border: solid 1px #37B24D;')
  })
})

describe('Alert custom', () => {
  it('should render component with custom styles if colors is custom', () => {
    const { styles } = renderWithTheme(
      <Alert color="custom" customIcon="outlined-default-mockup" customBackgroundColor="#ff0000">Lorem, ipsum dolor sit amet consectetur adipisicing elit.</Alert>
    )

    expect(styles.toString()).toContain('background: #ff0000')
    expect(screen.getByTestId('icon-outlined-default-mockup')).toBeInTheDocument()
  })

  it('should render component with custom styles and without icon', () => {
    const { styles } = renderWithTheme(
      <Alert color="custom" customBackgroundColor="#ff0000">Lorem, ipsum dolor sit amet consectetur adipisicing elit.</Alert>
    )

    expect(styles.toString()).toContain('background: #ff0000')
    expect(screen.queryByTestId('icon-outlined-default-mockup')).not.toBeInTheDocument()
  })
})
