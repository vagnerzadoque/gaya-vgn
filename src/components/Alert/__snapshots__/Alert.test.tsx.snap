// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert component should render component with info styles if color is not defined 1`] = `
[
  ".icon-0-2-5 {
  font-family: natds-icons;
  user-select: none;
  pointer-events: none;
}
.icon-d0-0-2-6 {
  color: #363636;
  font-size: 24px;
}
.wrapper-0-2-1 {}
.content-0-2-2 {}
.content-0-2-2 > h6 {
  margin: 0 0 8px 0;
  font-size: 20px;
}
.wrapper-d0-0-2-3 {
  gap: 8px;
  border: none;
  display: flex;
  padding: 16px;
  background: #D0EBFF;
  font-family: "Roboto";
  border-radius: 4px;
}
.content-d1-0-2-4 {
  color: #363636;
}",
  <div>
    <div
      class=" wrapper-0-2-1 wrapper-d0-0-2-3 Alert-gaya"
    >
      <i
        aria-hidden="true"
        class=" icon-0-2-5 icon-d0-0-2-6 natds-icons natds-icons-outlined-alert-info Icon-gaya"
        data-testid="icon-outlined-alert-info"
        role="img"
      />
      <div
        class="content-0-2-2 content-d1-0-2-4"
      >
        <h6>
          Title
        </h6>
        Lorem, ipsum dolor sit amet consectetur adipisicing elit.
      </div>
    </div>
  </div>,
]
`;
